src/start/AuthScreen.js

```diff
@@ -227,12 +240,66 @@ class AuthScreen extends PureComponent<Props> {
     this.props.dispatch(navigateToPassword(serverSettings.require_email_format_usernames));
   };
 
-  handleAuth = (method: AuthenticationMethodDetails) => {
+  handleNativeAppleAuth = async () => {
+    const state = await webAuth.generateRandomToken();
+    const credential: AppleAuthenticationCredential = await AppleAuthentication.signInAsync({
+      state,
+      requestedScopes: [
+        AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
+        AppleAuthentication.AppleAuthenticationScope.EMAIL,
+      ],
+    });
+    if (credential.state !== state) {
+      throw new Error('`state` mismatch');
+    }
+
+    otp = await webAuth.generateOtp();
+
+    const params = encodeParamsForUrl({
+      mobile_flow_otp: otp,
+      native_flow: true,
+      id_token: credential.identityToken,
+    });
+
+    openLink(`${this.props.realm}/complete/apple/?${params}`);
+
+    // Currently, the rest is handled with the `zulip://` redirect,
+    // same as in the web flow.
+    //
+    // TODO: Maybe have an endpoint we can just send a request to,
+    // with `fetch`, and get the API key right away, without ever
+    // having to open the browser.
+  };
+
+  handleAuth = async (method: AuthenticationMethodDetails) => {
     const { action } = method;
+
+    let shouldUseNativeAppleFlow: boolean = false;
+    if (
+      Platform.OS === 'ios'
+      && method.name === 'apple'
+      && (await AppleAuthentication.isAvailableAsync())
+    ) {
+      let host: string | void;
+      try {
+        // Check that the realm we're actually sending requests to,
+        // which is basically the URL the user entered on the first
+        // screen, is Kandra-hosted.
+        host = new WhatwgURL(this.props.realm).host;
+      } catch (e) {
+        // `this.props.realm` invalid. TODO: Check this much sooner.
+      }
+      if (config.kandraHostedDomains.includes(host)) {
+        shouldUseNativeAppleFlow = true;
+      }
+    }
```

Comment:
This is enough code it should probably get moved to its own little helper method. Then call like `} else if (this.shouldUseNativeAppleFlow(method)) {`.

Oh, or even
```
  } else if (method.name === 'apple'
    && this.canUseNativeAppleFlow()) {
```

That's nice because it separates the part that's about this auth method from the more-complicated rest.